pkg/util/globallimiter/limiter.go

```diff
@@ -0,0 +1,73 @@
+package globallimiter
+
+import (
+	"sync"
+	"time"
+)
+
+type GlobalLimiter interface {
+	Dec(key string)
+	Inc(key string) bool
+}
+
+func NewGlobalLimiter(limit int, activeKeys map[string]time.Time, timeout time.Duration) GlobalLimiter {
+	if activeKeys == nil {
+		activeKeys = map[string]time.Time{}
+	}
+	limiter := &realGlobalLimiter{
+		limit:      limit,
+		activeKeys: activeKeys,
+	}
+	go limiter.gcDeadKey(timeout)
+	return limiter
+}
+
+type realGlobalLimiter struct {
+	limit      int
+	lock       sync.Mutex
+	activeKeys map[string]time.Time
+}
+
+func (g *realGlobalLimiter) Dec(key string) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+	delete(g.activeKeys, key)
+}
+
+func (g *realGlobalLimiter) Inc(key string) bool {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+	if _, ok := g.activeKeys[key]; ok {
+		g.activeKeys[key] = time.Now()
+		return true
+	}
+	if len(g.activeKeys) < g.limit {
+		g.activeKeys[key] = time.Now()
+		return true
+	}
+	return false
+}
+
+func (g *realGlobalLimiter) gcDeadKey(timeout time.Duration) {
```

Comment:
plz comment about why we need gc keys, it that means the pulling job can be counted in the limit for almost timeout time?

Reply:
added comments.